File: C/Thread/5.8.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <stdlib.h>

sem_t sem1, sem2; // 2 sémaphores pour synchronisation

void* thread1(void* arg) 
{
    for(int i=0; i<10; i++) 
    {
        printf("T1 : traitement iteration %d\n", i+1);
        sleep(2); // traitement de 2 sec
        sem_post(&sem1); // signal que T1 a fini
        sem_wait(&sem2); // attend T2
    }
    return NULL;
}

void* thread2(void* arg) 
{
    for(int i=0; i<10; i++) {
        sem_wait(&sem1); // attend T1
        int t = rand()%6 + 4; // traitement aléatoire 4-9 sec
        printf("T2 : traitement iteration %d pour %d sec\n", i+1, t);
        sleep(t);
        sem_post(&sem2); // signal que T2 a fini
    }
    return NULL;
}

int main() {
    pthread_t t1, t2;

    sem_init(&sem1, 0, 0); // initialisation des sémaphores à 0
    sem_init(&sem2, 0, 0);

    pthread_create(&t1, NULL, thread1, NULL);
    pthread_create(&t2, NULL, thread2, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    sem_destroy(&sem1);
    sem_destroy(&sem2);

    return 0;
}

File: Python/Thread/5.8.py
import threading
import time
import random

sem1 = threading.Semaphore(0)
sem2 = threading.Semaphore(0)

def T1():
    for i in range(10):
        print(f"T1 : traitement iteration {i+1}")
        time.sleep(2)
        sem1.release()  # signal T1 fini
        sem2.acquire()  # attend T2

def T2():
    for i in range(10):
        sem1.acquire()  # attend T1
        t = random.randint(4,9)
        print(f"T2 : traitement iteration {i+1} pendant {t} sec")
        time.sleep(t)
        sem2.release()  # signal T2 fini

t1 = threading.Thread(target=T1)
t2 = threading.Thread(target=T2)

t1.start()
t2.start()

t1.join()
t2.join()